### DISCUSSION TITLE
RP6502 build problem with RIA and xram overflow.

### DISCUSSION THREAD

**Discussion by markrvmurray** [picocomputer/community#88] [Q&A]
Hi *,

I'm having a build problem on macOS Sonoma with the `arm-none-eabi` GCC from homebrew:

```
% mkdir build && cd build && cmake ..
% make -j4
[  0%] Performing build step for 'ELF2UF2Build'
:
[ 78%] Built target rp6502_vga
/Applications/ArmGNUToolchain/13.2.Rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: rp6502_ria.elf section `.bss' will not fit in region `RAM'
/Applications/ArmGNUToolchain/13.2.Rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: region RAM overflowed
/Applications/ArmGNUToolchain/13.2.Rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: section .stack1_dummy VMA [20040000,200407ff] overlaps section .bss VMA [2002f718,200459b7]
/Applications/ArmGNUToolchain/13.2.Rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: region `RAM' overflowed by 22968 bytes
collect2: error: ld returned 1 exit status
make[2]: *** [src/rp6502_ria.elf] Error 1
make[1]: *** [src/CMakeFiles/rp6502_ria.dir/all] Error 2
make: *** [all] Error 2
```

If I reduce the size of the `xram` array in `src/ria/sys/mem.[ch]` to e.g. `0x1000`, then the above error goes away. No point in trying to run it, though!

Could some kind soul with a working build please make your `rp6502_ria.elf.map` file from the build available?

Thanks!

M
--

**Comment by rumbledethumps** [comment]
[build.zip](https://github.com/picocomputer/community/files/14377283/build.zip)

**Reply by markrvmurray** [reply]
Thanks for the logs!

The problem is that with my later GCC-13 (you have GCC-11), the `.text` segment was too big. By starting the build with `cmake -DCMAKE_BUILD_TYPE=MinSizeRel -GNinja ..`, I got a smaller `.text` segment which doesn't annoy the linker.

**Reply by markrvmurray** [reply]
Diffing the logs was a bit of a chore, but I bodged up `sed` horror that changed most of the gratuitous differences like paths, long numbers beginning with `0x00000000` etc to something that `diff` could ignore. After that, seeing that it was a simple increase in size of a lot of library parts. I didn't deep-dive into this, but I got the impression that the floating-point stuff was a lot bigger, particularly when it came to I/O and string conversions.

**Reply by markrvmurray** [reply]
I installed an Ubuntu 22.* VM for testing, and its GCC-12 built with no issues and no need to force minimum size. Seems GCC-13 is possibly doing overly aggressive optimisations such as loop unrolling.

**Comment by rumbledethumps** [comment]
`RAM' overflowed by 22968 bytes from inlining or unrolling? A few kb sure, but a bump in compiler versions would not have been my first guess.

**Reply by markrvmurray** [reply]
Looks like it could well be a combination of compiler+optimizations. I can do `MinSizerel` builds with no problem. This build optimises with `-Os -DNDEBUG`. Builds of `Debug`/`-Og -g`, `Release`/`-O3 -DNDEBUG` and `RelWithDebugInfo`/`-O2 -g -DNDEBUG` all fail with varying amounts of `.text` segment overflow, in the range of about 9k to 90k.

I'm attaching my diff, which is the difference between your and my `Release` `rp6502_ria.elf.map` logs, with a klugy `sed` script used to clean up gratuitous path and file name differences. Note that my build fails, so look for the string `RAM`, and disregard the obvious extra stuff from a few paragraphs below that onwards.

It's a big diff, so figuring out the problem is tedious, but I still reckon it's down to optimization, where none or too much both result in larger binaries overall. 

[yours_mine.patch](https://github.com/picocomputer/community/files/14396328/yours_mine.patch)

**Reply by markrvmurray** [reply]
The GCC optimisations are documented here: https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html. These docs specifically say that "-Os enables all -O2 optimizations except those that often increase code size:".
